bip-0174.mediawiki

```diff
@@ -305,6 +305,53 @@ Signers do not need to sign for all possible input types. For example, a signer
 
 A single entity is likely to be both a Signer and an Updater as it can update a PSBT with necessary information prior to signing it.
 
+====Data Signers Check For====
+
+For a Signer to only produce valid signatures for what it expects to sign, it must check that the following conditions are true:
+
+* If a non-witness UTXO is provided, it's hash must match the hash specified in the prevout
+* If a witness UTXO is provided, a witness signature is being produced
+* If a redeemScript is provided, the scriptPubKey must be for that redeemScript
+* If a witnessScript is provided, the scriptPubKey or the redeemScript must be for that witnessScript
+
+=====Simple Signer Algorithm=====
+
+A simple signer can use the following algorithm to determine what and how to sign
+
+<pre>
+sign_witness(script_code, i):
+    for key in psbt.inputs[i].keys:
+        if IsMine(key):
+            sign(witness_sighash(script_code, i, input))
+
+sign_non_witness(script_code, i):
+    for key in psbt.inputs[i].keys:
+        if IsMine(key):
+            sign(non_witness_sighash(script_code, i, input))
+
+for input,i in enumerate(psbt.inputs):
+    if non_witness_utxo.exists:
+        assert(sha256d(non_witness_utxo) == prevhash)
+        if redeemScript.exists:
+            assert(non_witness_utxo.scriptPubKey == P2SH(redeemScript))
+            sign_non_witness(redeemScript)
+        else:
+            sign_non_witness(non_witness_utxo.scriptPubKey)
+    else if witness_utxo.exists:
+        if redeemScript.exists:
+            assert(witness_utxos.scriptPubKey == P2SH(redeemScript))
+            script = redeemScript
+        else:
+            script = witness_utxo.scriptPubKey
+        if IsP2WPKH(script):
+            sign_witness(P2PKH(script))
```

Comment:
You can't really convert a script to P2PKH. What about `P2PKH(script[2:])`?

Reply:
Done